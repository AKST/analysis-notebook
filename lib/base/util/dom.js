/**
 * @param {Node} parent
 * @param {Node} child
 */
export function insertAtStart(parent, child) {
  if (parent.firstChild) {
    parent.insertBefore(child, parent.firstChild);
  } else {
    parent.appendChild(child);
  }
}

/**
 * @param {HTMLElement | undefined} element
 * @param {number | boolean | string | null | undefined} value
 */
export function setDisable(element, value) {
  if (element == null) return;
  switch (value) {
    case undefined:
    case null:
    case '':
    case false:
    case 0:
      element.removeAttribute('disabled');
      break;
    default:
      element.setAttribute('disabled', 'disabled');
  }
}
