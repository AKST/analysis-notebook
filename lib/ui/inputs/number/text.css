:host {
  display: grid;
  grid-template-columns: 1fr;
  align-content: center;
}

.container {

}

input {
  width: 100%;
  border: var(--border-inset-grey);
  outline: none;
  border-radius: 0;
  background: white;
  color: black;
  font-family: monospace;
  line-height: 1;
  box-sizing: border-box;

  &.small {
    padding: 4px 8px;
    font-size: 10px;
    border-width: 1.5px;

    &.hide-border {
      padding: 2px 2px;
    }
  }

  &.medium {
    padding: 4px 8px;
    font-size: 12px;
    border-width: 3px;
  }

  &.hide-border {
    border: none;
    color: inherit;
    background: none;
  }

  &:focus {
    outline: none;
    border-color: #0084ff;
  }

  &:invalid {
    border-color: red;
    color: red;
  }

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
}
