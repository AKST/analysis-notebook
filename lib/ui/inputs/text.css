:host {
  display: block;
}

input {
  width: 100%;
  padding: 4px 8px;
  border: var(--border-inset-grey);
  border-radius: 0;
  background: white;
  color: black;
  font-family: monospace;
  font-size: 12px;
  box-sizing: border-box;

  &:focus {
    outline: none;
    border-color: #0084ff;
    box-shadow: 0 0 0 2px rgba(0, 132, 255, 0.2);
  }

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
}
