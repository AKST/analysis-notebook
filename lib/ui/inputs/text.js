import { generateCSS, updateOn } from '../../base/render_ui/index.js';

function getStyle() {
  return generateCSS({
    ':host': {
      display: 'block',
    },

    'input': {
      width: '100%',
      padding: [4, 8],
      border: 'var(--border-inset-grey)',
      borderRadius: 0,
      background: 'white',
      color: 'black',
      fontFamily: 'monospace',
      fontSize: 12,
      boxSizing: 'border-box',
    },

    'input:focus': {
      outline: 'none',
      borderColor: '#0084ff',
      boxShadow: '0 0 0 2px rgba(0, 132, 255, 0.2)',
    },

    'input:disabled': {
      opacity: 0.5,
      cursor: 'not-allowed',
    }
  });
}

/**
 * Text Input Web Component
 * Based on RFC 006 - reusable input for use in knobs and other components
 */
export class TextInput extends HTMLElement {
  static observedAttributes = ['value', 'placeholder', 'disabled'];

  /** @type {string | undefined} */
  placeholder;

  disabled = false;
  #value = '';

  /** @type {HTMLInputElement | null} */
  #input = null;

  /**
   * @param {string} [value='']
   * @param {string} [placeholder]
   * @param {boolean} [disabled=false]
   */
  constructor(value, placeholder, disabled) {
    super();
    this.attachShadow({ mode: 'open' });
    this.#value = value ?? '';
    this.disabled = disabled ?? false;
    this.placeholder = placeholder;
  }

  connectedCallback() {
    this.render();
    this.#input = /** @type {HTMLInputElement | null} */ (this.shadowRoot?.querySelector('input'));
    this.setupEventListeners();
  }

  /**
   * @param {string} name
   * @param {string | null} oldValue
   * @param {string | null} newValue
   */
  attributeChangedCallback(name, oldValue, newValue) {
    if (oldValue !== newValue) {
      this.updateAttribute(name, newValue);
      this.render();
    }
  }

  setupEventListeners() {
    this.shadowRoot?.addEventListener('input', (e) => {
      if (!(e.target instanceof HTMLInputElement)) return
      this.#value = e.target.value;
      this.dispatchEvent(new CustomEvent('change', {
        detail: { value: this.#value },
        bubbles: true
      }));
    });

    this.shadowRoot?.addEventListener('keydown', (e) => {
      if (!(e.target instanceof HTMLInputElement)) return
      if (!(e instanceof KeyboardEvent && e.key === 'Enter')) return;
      this.dispatchEvent(new CustomEvent('submit', {
        detail: { value: this.#value },
        bubbles: true
      }));
    });
  }

  /**
   * @param {string} value
   */
  setValue(value) {
    this.#value = value || '';
    if (this.#input) {
      this.#input.value = value;
    }
    this.render();
  }

  getValue() {
    return this.#value;
  }

  /**
   * @param {string} name
   * @param {any} value
   */
  updateAttribute(name, value) {
    switch (name) {
      case 'value':
        this.#value = value || '';
        if (this.#input) {
          this.#input.value = value;
        }
        break;
      case 'placeholder':
        this.placeholder = value;
        break;
      case 'disabled':
        this.disabled = value === 'disabled' || value === true;
    }
  }

  render() {
    updateOn(this.shadowRoot, [
      ['style', {}, [getStyle()]],
      ['input', {
        type: 'text',
        value: this.#value,
        placeholder: this.placeholder,
        disabled: this.disabled ? 'disabled' : undefined,
      }, []]
    ]);
  }
}

customElements.define('text-input', TextInput);
