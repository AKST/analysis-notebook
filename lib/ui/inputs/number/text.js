/**
 * @import { El } from '../../../base/dom_ui/index.js';
 *
 * @typedef {(
 *   | { valid: true, value: number, text: string }
 *   | { valid: false, value: number | undefined, text: string }
 * )} NumberTextChangeEventDetails
 *
 * @typedef {'small' | 'medium'} ScaleAppearance
 */
import { Unreachable } from '../../../base/util/type.js';
import { setDisable } from '../../../base/util/dom.js';
import { eventFilterByElem } from '../../../base/util/dom.js';
import { render, updateOn } from '../../../base/dom_ui/index.js';
import { SharedStyleSheet } from '../../../base/dom_ui/shared_style_sheet.js';

/**
 * @extends {CustomEvent<NumberTextChangeEventDetails>}
 */
export class NumberTextChangeEvent extends CustomEvent {
  /**
   * @param {NumberTextChangeEventDetails} detail
   * @param {Omit<CustomEventInit<unknown>, 'detail'>} [init]
   */
  constructor(detail, init) {
    super('change', {
      bubbles: true,
      composed: true,
      ...init,
      detail,
    });
  }
}

export class NumberTextInput extends HTMLElement {
  static sheet = new SharedStyleSheet('./lib/ui/inputs/number/text.css');
  static get observedAttributes() {
    return ['disabled', 'value', 'scale', 'showBorder'];
  }

  /** @type {ShadowRoot} */ #shadow;
  /** @type {HTMLInputElement | undefined} */ #input;
  /** @type {'float' | 'int'} */ #numberKind;
  /** @type {string} */ #text;
  /** @type {number} */ #min;
  /** @type {number} */ #max;
  /** @type {number} */ #inc;
  /** @type {ScaleAppearance} */ #appearanceScale;
  /** @type {boolean} */ #appearanceShowBorder;

  /**
   * @param {string} [text]
   * @param {number} [min]
   * @param {number} [max]
   * @param {number} [inc]
   * @param {'float' | 'int'} [numberKind]
   * @param {ScaleAppearance} [scale]
   * @param {boolean} [showBorder]
   */
  constructor(
      text='0',
      min = -Infinity,
      max = Infinity,
      inc = 1,
      numberKind = 'float',
      scale = 'medium',
      showBorder = true,
  ) {
    super();
    this.#shadow = this.attachShadow({ mode: 'open' });
    this.#input = undefined;
    this.#text = text;
    this.#min = min;
    this.#max = max;
    this.#inc = inc;
    this.#numberKind = numberKind;
    this.#appearanceScale = scale;
    this.#appearanceShowBorder = showBorder;
  }

  /**
   * @param {number} value
   * @param {number} [min]
   * @param {number} [max]
   */
  static createFromNumber(value, min, max) {
    return new NumberTextInput(`${value}`, min, max);
  }

  connectedCallback() {
    NumberTextInput.sheet.loadSheet(sheet => this.#shadow.adoptedStyleSheets = [sheet]);
    const input = this.#render();

    this.#shadow.addEventListener('input', (
      eventFilterByElem(InputEvent, input, this.onChange)
    ));

    this.#shadow.addEventListener('keydown', (
      eventFilterByElem(KeyboardEvent, input, this.onKeyPress)
    ));
  }

  /**
   * @param {HTMLInputElement} element
   * @param {InputEvent} _event
   */
  onChange = (element, _event) => {
    this.#internalSetValue(element.value);
  };

  /**
   * @param {HTMLInputElement} element
   * @param {KeyboardEvent} event
   */
  onKeyPress = (element, event) => {
    let diff = 0;
    const m = event.shiftKey ? 10 : 1;
    switch (event.key) {
      case 'ArrowUp':
        diff = this.#inc * m;
        break;
      case 'ArrowDown':
        diff = -(this.#inc * m);
        break;
      default:
        return;
    }
    event.preventDefault();
    if (this.value == null) return;

    const update = this.value + diff;
    element.value = update+'';
    this.#internalSetValue(update+'');
  };

  /**
   * @returns {HTMLInputElement}
   */
  #render() {
    const input = this.#input = render(['input', {
      type: 'text',
      value: this.#text,
      disabled: this.hasAttribute('disabled') || null,
    }]);
    this.setScaleAppearance(this.#appearanceScale);
    this.setShowBorderAppearance(this.#appearanceShowBorder);
    updateOn(this.shadowRoot, [input]);
    return input;
  }

  /**
   * @param {string} name
   * @param {any} oldValue
   * @param {any} newValue
   */
  attributeChangedCallback(name, oldValue, newValue) {
    if (oldValue == newValue) return;

    switch (name) {
      case 'disabled':
        setDisable(this.#input, newValue);
        break;

      case 'value':
        this.#internalSetValue(newValue);
        break;

      case 'scale':
        if (!['small', 'medium'].includes(newValue)) break;
        this.setScaleAppearance(newValue);
        break;

      case 'show-border':
        this.setShowBorderAppearance(oldValue === false || oldValue !== 'false');
        break;
    }
  }

  /** @param {string} text */
  boardCastSet(text) {
    this.#internalSetValue(text);
  }

  get valid() {
    const value = this.value;
    if (value == null) return false;
    if (Number.isNaN(value)) return false;
    if (value > this.#max) return false;
    if (value < this.#min) return false;
    return true;
  }

  get text() { return this.#text; }
  set text(value) {
    this.#text = value;

    if (this.#input) {
      this.#input.value = value;
    }
  }

  /**
   * @type {number | undefined}
   */
  get value() {
    try {
      switch (this.#numberKind) {
        case 'int':
          return parseInt(this.#text, 10);

        case 'float':
          return parseFloat(this.#text);

        default:
          throw new Unreachable(this.#numberKind);
      }
    } catch (e) {
      console.info(e);
      return undefined;
    }
  }

  /**
   * @param {string} text
   */
  #internalSetValue(text) {
    this.#text = text;
    const valid = this.valid, value = this.value;
    if (!valid) {
      this.#input?.setCustomValidity("Field is invalid.");
      this.dispatchEvent(new NumberTextChangeEvent({
        text,
        valid: false,
        value,
      }));
    } else {
      this.#input?.setCustomValidity("");
      this.dispatchEvent(new NumberTextChangeEvent({
        text,
        valid: true,
        value: /** @type {any} */ (value),
      }));
    }
  }

  /** @param {ScaleAppearance} scale */
  setScaleAppearance(scale) {
    this.#appearanceScale = scale;
    if (this.#input == null) return;

    switch (scale) {
      case 'small':
        this.#input.classList.add('small');
        this.#input.classList.remove('medium');
        break;

      case 'medium':
        this.#input.classList.remove('small');
        this.#input.classList.add('medium');
        break;
      default:
        throw new Unreachable(scale);
    }
  }

  /** @param {boolean} showBorder */
  setShowBorderAppearance(showBorder) {
    this.#appearanceShowBorder = showBorder;
    if (this.#input == null) return;
    if (showBorder) {
      this.#input.classList.remove('hide-border');
    } else {
      this.#input.classList.add('hide-border');
    }
  }
}

customElements.define('input-number-text', NumberTextInput);
