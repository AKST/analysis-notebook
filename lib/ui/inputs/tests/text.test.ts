import { expect, describe, it, beforeEach, vi } from 'vitest';
import userEvent from '@testing-library/user-event';
import { getByRole, getByDisplayValue, getByPlaceholderText } from '@testing-library/dom';
import { TextInput } from '../text.js';

describe('TextInput', () => {
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    document.body.innerHTML = '';
    user = userEvent.setup();
  });

  const create = (value?: string, placeholder?: string, disabled?: boolean) => {
    const textInput = new TextInput(value, placeholder, disabled);
    document.body.appendChild(textInput);
    return textInput;
  };

  it('renders text input correctly', () => {
    const textInput = create('Hello world', 'Enter text...');
    expect(textInput.shadowRoot?.innerHTML).toMatchSnapshot();
  });

  it('responds to user typing', async () => {
    const textInput = create('', 'Type here...');
    const changeHandler = vi.fn();
    textInput.addEventListener('change', changeHandler);

    const input = getByPlaceholderText(textInput.shadowRoot as any, 'Type here...');
    await user.type(input, 'Hello world');

    expect(changeHandler).toHaveBeenCalled();
    expect(textInput.getValue()).toBe('Hello world');
  });

  it('responds to user pressing Enter', async () => {
    const textInput = create('Test value');
    const submitHandler = vi.fn();
    textInput.addEventListener('submit', submitHandler);

    const input = getByDisplayValue(textInput.shadowRoot as any, 'Test value');
    await user.type(input, '{Enter}');

    expect(submitHandler).toHaveBeenCalledOnce();
    expect(submitHandler.mock.calls[0][0].detail.value).toBe('Test value');
  });
});
