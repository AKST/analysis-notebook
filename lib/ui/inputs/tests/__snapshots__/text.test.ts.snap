// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextInput > renders text input correctly 1`] = `
"<style>:host {
  display: block;
}

input {
  width: 100%;
  padding: 4px 8px;
  border: var(--border-inset-grey);
  border-radius: 0;
  background: white;
  color: black;
  font-family: monospace;
  font-size: 12px;
  box-sizing: border-box;
}

input:focus {
  outline: none;
  border-color: #0084ff;
  box-shadow: 0 0 0 2px rgba(0, 132, 255, 0.2);
}

input:disabled {
  opacity: 0.5px;
  cursor: not-allowed;
}

</style><input type="text" value="Hello world" placeholder="Enter text...">"
`;
